text/0031-configuration.md

```diff
@@ -0,0 +1,287 @@
+
+# Motivation
+
+This RFC contains a set of recommendations regarding configuration management in order to ensure consistency across the
+different Bee crates.
+
+# Detailed design
+
+## Libraries
+
+- [serde](https://github.com/serde-rs/serde): the go-to framework for **ser**ializing and **de**serializing Rust data
+  structures efficiently and generically;
+- [toml-rs](https://github.com/alexcrichton/toml-rs): a TOML encoding/decoding library for Rust with
+  serialization/deserialization on top of `serde`;
+
+## Recommendations
+
+With the following recommendations, one can:
+- read a configuration builder from a file;
+- write a configuration to a file;
+- manually override a configuration builder;
+- construct the configuration;
+
+### Configuration Builder
+
+The [Builder](https://rust-lang.github.io/api-guidelines/type-safety.html#c-builder) pattern is very common in Rust when
+it comes to constructing complex objects like configurations.
+
+A configuration builder type should:
+- be named `*ConfigBuilder`;
+- derive the following traits;
+    - `Default` to easily implement the `new` method as convention;
+    - `Deserialize`, from `serde`, to deserialize from a configuration file;
+- provide a `new` method;
+- have `Option` fields;
+    - if there is a configuration file, the fields should have the same names as the keys;
+- provide setters to set/override the fields;
+- provide a `finish` method constructing the actual configuration object;
+- have default values defined as `const` and set with `Option::unwrap_or`/`Option::unwrap_or_else`;
+
+Here is a small example fitting all these requirements.
+
+`config.toml`
+```toml
+[snapshot]
+meta_file_path  = "./data/snapshot/mainnet.snapshot.meta"
+state_file_path = "./data/snapshot/mainnet.snapshot.state"
+```
+
+`config.rs`
+```rust
+const DEFAULT_META_FILE_PATH: &str = "./data/snapshot/mainnet.snapshot.meta";
+const DEFAULT_STATE_FILE_PATH: &str = "./data/snapshot/mainnet.snapshot.state";
+
+#[derive(Default, Deserialize)]
+pub struct SnapshotConfigBuilder {
+    meta_file_path: Option<String>,
+    state_file_path: Option<String>,
+}
+
+impl SnapshotConfigBuilder {
+    pub fn new() -> Self {
+        Self::default()
+    }
+
+    pub fn meta_file_path(mut self, meta_file_path: String) -> Self {
+        self.meta_file_path.replace(meta_file_path);
+        self
+    }
+
+    pub fn state_file_path(mut self, state_file_path: String) -> Self {
+        self.state_file_path.replace(state_file_path);
+        self
+    }
+
+    pub fn finish(self) -> SnapshotConfig {
+        SnapshotConfig {
+            meta_file_path: self.meta_file_path.unwrap_or(DEFAULT_META_FILE_PATH.to_string()),
+            state_file_path: self.state_file_path.unwrap_or(DEFAULT_STATE_FILE_PATH.to_string()),
+        }
+    }
+}
+```
+
+### Configuration
+
+A configuration type should:
+- be named `*Config`;
+- derive the following traits;
+  - `Clone` to be able to provide ownership to different components;
```

Comment:
^ that is slightly confusing, because the way you write it one could ask: "Why is he not using Rc/Arc then?" Would be nice if you'ld explain why opting for `Clone` here ;)

Reply:
Should I add a `unique` before ownership ? It feels weird to have a justify not providing shared ownership in this context. What would you say ?

Reply:
`Since all configuration objects are most probably aggregated in a common configuration object after reading from a file, `Clone` is needed to give components their unique ownership of their configuration`.  Something like that would be fine ?

Reply:
perfect :+1: